### ISSUE TITLE
How to release GPU memory after sess.close()?

### ISSUE THREAD

**Issue by githubgsq** [tensorflow/tensorflow#19731]
hi, all:
     I'm training models iteratively. After each model trained, I run sess.close() and recreate a new session to run a new training process. But it seems that the GPU memory was not relseased and it's increasing constantly. 
I tried tf.reset_default_graph() before run session also typed gc.collect() after sess.close(), but takes no effect. 
     How could I release GPU memory timely to avoid OOM error please? 
     Thanks!

**Comment by githubgsq** [comment]
hi, @JaeDukSeo 
thanks for your kind advices and I tried them
1. I initialized session with config as you said.
    config = tf.ConfigProto()
    config.gpu_options.allow_growth=True
    sess = tf.Session(config=config)
2. I didn't set 'per_process_gpu_memory_fraction', because I have only one process and total gpu memory can be used
3. I tried sess.reset(), but errors occured, which reset() takes at least 1 argument (0 given). https://www.tensorflow.org/api_docs/python/tf/Session#reset shows that reset() is implemented for distributed session.  I didn't use distributed session also did not set target arg here.

Why the gpu memory usage is still lingering after sess.close() and del graph?
Is there any other advice for releasing resources?

**Comment by githubgsq** [comment]
hi,all
    I tried method referred here:https://github.com/tensorflow/tensorflow/issues/17048
    and the memory could be released now!
    Thanks a lot~

**Comment by JaeDukSeo** [comment]
haha I am glad! and my bad the commend was actually 
tf.reset_default_graph()
<img width="410" alt="screen shot 2018-06-04 at 8 34 06 am" src="https://user-images.githubusercontent.com/22832406/40917823-8f7fc1a6-67d2-11e8-92a9-19191a26319d.png">

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by saxenarohan97** [comment]
I have a similar problem. I have a couple of questions in this regard:

@githubgsq when you mention about the method from [#17048](https://github.com/tensorflow/tensorflow/issues/17048), do you mean moving your TensorFlow session code into a subprocess? So when the subprocess exits, the GPU memory is released?

@JaeDukSeo You mention setting `allow_growth=True`, but if my model is very large and a large amount of memory is allocated to TensorFlow even after `allow_growth=True`, then it will not be deallocated, right? (Since the [docs](https://www.tensorflow.org/programmers_guide/using_gpu#allowing_gpu_memory_growth) say that "Note that we do not release memory, since that can lead to even worse memory fragmentation").

Thanks in advance.

**Comment by githubgsq** [comment]
@saxenarohan97 yes, a subprocess run the session code. I also called tf.reset_default_graph() before the subprocess executed.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@JaeDukSeo do you happen to have an answer for @saxenarohan97 ?

**Comment by JaeDukSeo** [comment]
@cy89  I agree with @saxenarohan97 if allow_growth is set to true I don't think it automatically deallocates GPU

**Comment by cy89** [comment]
@JaeDukSeo thanks for your reply!
I'll close, as it looks like this thread has answers to all open questions.

**Comment by TanLingxiao** [comment]
I use [numba ](http://numba.pydata.org/numba-doc/0.13/CUDADevice.html)to release the gpu. With tensorflow I can not find a effect method.

**Comment by nikolayvoronchikhin** [comment]
@TanLingxiao were you able to find any other method? numba is a great way with the drawback being that once you run cuda.close(), you can no longer run your process again in the same process/session. Was hoping that tensorflow has config option to free GPU Memory after the processing ends.

**Comment by p890040** [comment]
> Hi I'm new to GitHub and have been trying to get Tensorflow running in Python.
> 
> In principle I have the thing up and running, but the GPU memory is not released, causing an OOM error at some point. These few lines already clutter the memory.
> 
> import tensorflow as tf
> sess = tf.Session()
> sess.close()
> 
> I've been googling the problem but so far I have only managed to solve it by either
> 
> * Clossing the console (and thus losing all varibales which is what I am actually trying to avoid)
> * Using numba to kill the cuda device (however, this only works once) - see above. In that case my code looked something like this:
> 
> from numba import cuda
> cuda.select_device(0)
> #do tf stuff
> cuda.close()
> #the memory was released here!
> cuda.select_device(0)
> #to tf stuff -> caused an OOM
> (compare to https://numba.pydata.org/numba-doc/dev/cuda/device-management.html)
> 
> As mentioned above I would like to avoid killing the session and thus losing my varibales in memory (used to train a NN).
> 
> * I am aware that I can alocate only a fraction of the memory (cfg.gpu_options.per_process_gpu_memory_fraction = 0.1) or let the memory grow (cfg.gpu_options.allow_growth=True) and this both works fine, but afterwards I simply am unable to release the memory.
> * I have tried to put it into threads and pools following ideas and workaounds people suggested.
> * Also the K.session_close() and/or gc.collect() combination does not work.
> * I also downgraded tf to 1.9 and Keras to 2.1.9 (second not really relevant for the above code).
> * I have also upgraded my graphics card driver to the newest release (see below and note the memory which is not released after the call from above).
> 
> ![grafik](https://user-images.githubusercontent.com/43452440/45842749-e6e0e600-bd1d-11e8-8023-387c219b1bbf.png)
> 
> I'd be very thankful for any suggestions what to do with the code snippet from above to ensure that the GPU memory is free in the end

Exactly same problem for me.
Very hope for some suggestion about this.

**Comment by anki-xyz** [comment]
Have the same issue hear; I can only fit a model once using Keras with TensorFlow backend, and the second time (with the very same model), it just crashes (OOM error).
Also appreciate suggestions here.

**Comment by pwais** [comment]
+1

**Comment by marinone94** [comment]
I have solved it by running the session in a separate Thread. When the session is completed, the memory used by such process is released, when the process is killed.  Remember to save your session results on disk in the same method.

**Comment by pwais** [comment]
@marinone94 do you have code / proof?  I tried a thread too but the only thing that worked for me is to use a subprocess.   Others have seen similar: https://github.com/tensorflow/tensorflow/issues/20387 https://github.com/tensorflow/tensorflow/issues/15880

**Comment by p890040** [comment]
@yurmchg @marinone94 
You guys are right. I have tried this way, and it does work.
However, I think it' not very efficient method for program. Because it may cause many problem and inconvenient when we want to communicate with other process(ex: send or receive data).
Very Hope tensorflow official can provide the release gpu memory function, or just make it release after end this thread (not process).

**Comment by pwais** [comment]
@p890040 can you please post a code example?  I definitely tried python threads to no avail.

**Comment by jessequinn** [comment]
> I use [numba ](http://numba.pydata.org/numba-doc/0.13/CUDADevice.html)to release the gpu. With tensorflow I can not find a effect method.

thanks for this!

**Comment by mrgloom** [comment]
`numba` cause Segmentation fault.

Example:
https://stackoverflow.com/questions/58792739/tensorflow-model-wrapper-that-can-release-gpu-resources

So as I understand the only method so far is to run model load and predict in separate process or maybe pass same session to any used model clearing old graph before loading new model.

**Comment by saravanabalagi** [comment]
I tried the following things, but, none **guaranteed** to free the memory up
```
# didn't work for me
tf.reset_default_graph()
K.clear_session()
cuda.select_device(0); cuda.close()
model = get_new_model() # overwrite
model = None
del model
gc.collect()
``` 
Creating [separate processes](https://stackoverflow.com/a/58944126/3125070) always worked and **guaranteed** that the memory is freed up than the following. Further that helped me manage and allocate resources the way I wanted.
```
# works for me
process_train = multiprocessing.Process(train_model, args=...())
process_train.start()
process_train.join()
```

**Comment by pwais** [comment]
+1 @saravanabalagi subprocess works for me, but that is in many cases impractical.

**Comment by mirekphd** [comment]
Seeing this issue (and its many variations) closed, we see it is a design flaw and should move on. For new projects that need no JavaScript runtime I will always recommend PyTorch, which has a dedicated function for releasing GPU memory, `torch.cuda.empty_cache()`

Saves a lot of hacking around memory management and as a bonus the API proved more stable over time...

**Comment by leimao** [comment]
> tf.reset_default_graph()

This is being deprecated by TF 2.0 and there is no equivalence in TF 2.0.

**Comment by leimao** [comment]
Please reopen this issue.

**Comment by EKami** [comment]
Same issue here, please reopen the issue... TF 2.0 didn't solve anything, just made my code run slower...

**Comment by asis-shukla** [comment]
tf.reset_default_graph() not working in 2.1.
How to release GPU memory after model training in Eager execution mode.

**Comment by pwais** [comment]
@cy89 Please re-open.  Closing this issue was not helpful to the community.

**Comment by duduscript** [comment]
Same issue, please reopen the issue.

**Comment by corvin-quodai** [comment]
Same issue, please reopen.

**Comment by dd1923** [comment]
Same issue. Seems crazy that a framework like Tensorflow does not even have a one simple way to release the memory.

**Comment by asis-shukla** [comment]
@dd1923 Can you try this one. On TensorFlow 2.1 it seems work
`import tensorflow.keras.backend as K
K.clear_session()`

**Comment by dd1923** [comment]
@asis-shukla Tried it. Does not work. Still leaves all gpus full unfortunately.

**Comment by forkbabu** [comment]
@asis-shukla doesn't work.

**Comment by GF-Huang** [comment]
Not work tf-2.2

**Comment by yinghuang** [comment]
Someone please give a correct answer!

**Comment by maxvfischer** [comment]
> Someone please give a correct answer!

@yinghuang  As a lot of other people, I'm using subprocesses to release the memory, which works fine.

E.g.
```python
import multiprocessing

def run_inference_or_training(param1, param2, ...):
    ...

if __name__ == '__main__':
    p = multiprocessing.Process(
        target=run_inference_or_training,
        args=(param1, param2, )
    )
    p.start()
    p.join()  # Add this if you want to wait for the process to finish.
```
When the function `run_inference_or_training` is done, the process `p` will be closed down and the memory allocated inside `run_inference_or_training` will be released.

**Comment by newbiesitl** [comment]
There are ways to reset default graph in tf2 
1:

with tf.Graph().as_default():
    main()
2:
from tensorflow.python.framework import ops
ops.reset_default_graph()
sess = tf.InteractiveSession()

**Comment by mushaffarasyidridha** [comment]
as stated in the link, DO NOT use memory growth:
"Once the memory is allocated, it will not be released to avoid memory fragmentation. So even if your process starts on low memory, it will acquire more memory gradually but it will not release when it don’t need it."
https://medium.com/ibm-data-ai/memory-hygiene-with-tensorflow-during-model-training-and-deployment-for-inference-45cf49a15688
use First Option — Specifically Set The Memory.

**Comment by OtakuOW** [comment]
Hi, I have the same problem. I have a 12Gb RTX 4070 OC but I can't create because it tells me I have no memory. Not being an expert in programming, could you explain to me very easily how to remedy this problem? Photos are also welcome
